### DISCUSSION TITLE
question/implement-promise-prototype-finally

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#40] [Announcements]
# question/implement-promise-prototype-finally

北美前端面试攻略

https://us-fe.github.io/question/implement-promise-prototype-finally.html

**Comment by xiaoyu2er** [comment]
我们先看一下 finally 的定义 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#description

> + The finally() method is very similar to calling .then(onFinally, onFinally) however there are a couple of differences:
> + When creating a function inline, you can pass it once, instead of being forced to either declare it twice, or create a variable for it
> A finally callback will not receive any argument, since there's no reliable means of determining if the promise was fulfilled or rejected. This use case is for precisely when you do not care about the rejection reason, or the fulfillment value, and so there's no need to provide it. So for example:
>   - Unlike Promise.resolve(2).then(() => {}, () => {}) (which will be resolved with undefined), Promise.resolve(2).finally(() => {}) will be resolved with 2.
>   - Similarly, unlike Promise.reject(3).then(() => {}, () => {}) (which will be fulfilled with undefined), Promise.reject(3).finally(() => {}) will be rejected with 3.

个人觉得最重要的是 finally 的返回值，仍然是原 then or catch 的返回值
参考方案

https://stackoverflow.com/questions/53327711/how-to-add-a-polyfill-to-support-finally-in-edge

https://github.com/zloirock/core-js/blob/master/packages/core-js/modules/es.promise.finally.js

**Reply by xiaoyu2er** [reply]
```js
/**
 * @param {Promise<any>} promise
 * @param {() => void} onFinally
 * @returns {Promise<any>}
 */
function myFinally(promise, onFinally) {

  return promise.then(
    (val) => {
      return Promise.resolve(onFinally()).then(() => val);
    }
    , (e) => {
      return Promise.resolve(onFinally()).then(() => {
        throw e
      })
    })
}
```

**Comment by xiaoyu2er** [comment]
我们来看一下题目，还是非常triky的

```js
Promise.resolve(1)
.finally((data) => {
  console.log(data) // undefined // 不接受入参
  return Promise.reject('error') // 这里比较关键，因为 finally 抛出错误，会被下面的 catch 接住
})
.catch((error) => {
  console.log(error) // ”error“
  throw 'error2'
})
.finally((data) => {
  console.log(data) // undefined // 不接受入参
  return Promise.resolve(2).then(console.log) // 自己先打印 2，然后把上个 promise 的返回值原封不动透传，即 throw 'error2'
})
.then(console.log)
.catch(console.log) // 被这里接住 打印 "error"

```
答案 
```js
undefined
"error"
undefined
2
"error2"
```
